### DISCUSSION TITLE
Mongoose websocket disables interrupt or some logic on stm32 nucleo h723zg?

### DISCUSSION THREAD

**Discussion by PietroElectr** [cesanta/mongoose#2803] [Ask questions, get answers] [answered]
Good morning everyone,

I'm writing here because I'm facing a problem in my application. Let me explain it to you.
I'm developing a system using an STM32 Nucleo board, more precisely an h723zg.
The overall application has to:

Continuously sample data using the internal 16-bit ADC at a 10kHz frequency.
Once the buffer is full, send the data using a websocket connection.
The routine continues as long as the client requests to stop, thus maintaining a data stream.

The ADC is triggered by a timer every 100µs to match the 10kHz frequency constraint. Every time it is triggered, the sampled data is saved in an array.
The problem is that I noticed as soon as I call the mg_ws_send function, the ADC stops working and doesn't follow the input signal anymore (the input signal at the moment is a simple 60Hz or higher frequency sinusoid). This causes a noticeable jump, as shown in the attached picture.
![jump](https://github.com/cesanta/mongoose/assets/103196253/83cfd526-4299-4a22-90e5-74f9c774f523)

I attempted the following solutions:

Use TIMER+ADC+DMA, but it didn't work.
Use ADC+DMA, but it didn't work.
Change the priority of the interrupt using the STM32 environment, but it didn't work.
I can solve the problem by using external flash memory to store the data during sampling and then sending it all in one shot, or by using an external ADC. However, I would like to know if there is a function or a change I can make to solve the problem using only a software approach.
Sorry for the question, I'm kind of new at this. I would like to thank you all; the Mongoose library is amazing.
If you need further details, don't hesitate to ask me.

Code implementation in short:

waiting for client request
  if (ev == MG_EV_WS_MSG)
  {
	  while (packet_ready != 1); //waiting for ADC filling up the buffer to be sent

	  mg_ws_send(c, (char *)samples, sizeof(samples), WEBSOCKET_OP_BINARY);
	  HAL_GPIO_TogglePin(LED_YELL_GPIO_Port, LED_YELL_Pin);
	  packet_ready = 0;

  }

ADC routine:

void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *hadc)
{
	samples[i] = HAL_ADC_GetValue(&hadc1);
	i++; 
	
	if (i == 10000){
		i=0;
		packet_ready = 1;
	}
}

Wish you a good day

**Comment by scaprile** [comment, marked as answer]
I don't know what you are doing nor what you are using, but Mongoose does NOT fiddle with interrupts (except, of course, if you use the firmware upload facility).
Mongoose does not fiddle with some logic nor any logic, nowhere. It is as much hardware agnostic as it can be (except, of course, if you use the internal TCP/IP stack, that needs an Ethernet controller driver)

You are not following good practices and are blocking at an event handler. Please don't. Please see our [documentation](https://mongoose.ws/documentation/),  and follow the guidelines in our [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples).
Your problem is not with Mongoose, you need a properly architected application, decoupling things where needed
That's all I can say.
